<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Splitting a Count Matrix into Independent Folds • countsplit</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Splitting a Count Matrix into Independent Folds">
<meta property="og:description" content="Implements the count splitting methodology from Neufeld et al. (2022) &lt;doi:10.1093/biostatistics/kxac047&gt; and Neufeld et al. (2023) &lt;arXiv:2307.12985&gt;. Intended for turning a matrix of single-cell RNA sequencing counts, or similar count datasets, into independent folds that can be used for training/testing or cross validation. Assumes that the entries in the matrix are from a Poisson or a negative binomial distribution.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">countsplit</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">4.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/anna-neufeld/countsplit/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level2">
<h2 id="what-is-countsplit">What is countsplit?<a class="anchor" aria-label="anchor" href="#what-is-countsplit"></a>
</h2>
<p>The <code>countsplit</code> R package splits an integer-valued matrix into multiple folds of data with the same dimensions. These folds will be independent under certain modeling assumptions, and can thus be used for cross validation.</p>
<p>For tutorials associated with this package, please visit <a href="https://anna-neufeld.github.io/countsplit.tutorials/" class="external-link">https://anna-neufeld.github.io/countsplit.tutorials/</a>.</p>
<p>The motivation for this method in the setting where the data are Poisson distributed is described in Neufeld et al., 2022 <a href="https://arxiv.org/abs/2207.00554" class="external-link">(link to paper)</a> in the context of inference after latent variable estimation for single cell RNA sequencing data. Briefly, count splitting allows users to perform differential expression analysis to see which genes vary across estimated cell types (such as those obtained via clustering) or along an estimated cellular trajectory (pseudotime). Neufeld et al., 2023 <a href="https://arxiv.org/pdf/2307.12985.pdf" class="external-link">(link to preprint)</a> extends the method to the setting where the data follow a negative binomial distributed, and provides additional settings where count splitting is useful. For example, count splitting is useful broadly for evaluating low-rank representations of the data.</p>
</div>
<div class="section level2">
<h2 id="recent-package-updates">Recent package updates<a class="anchor" aria-label="anchor" href="#recent-package-updates"></a>
</h2>
<p>We recently sped up the performance of the package by re-implementing the main functions in C++. This is especially useful for real scRNA-seq datasets, which are quite large. We would like to acknowledge Mischko Heming (mheming.de) for implementing most of this speedup through a github contribution.</p>
<p>We have consolidated the functions in this package such that both Poisson and negative binomial thinning can be performed using the same function; <code>countsplit</code>. This function can also be used to create an arbitrary number of folds of data, rather than just a single train/test split. If you are a previous user of countsplit, please be sure to read the documentation to see our recent changes!</p>
<p>The vignettes and data associated with this package are stored in the associated ``countsplit.tutorials” package. To see the tutorials, please visit the updated tutorial website: <a href="https://anna-neufeld.github.io/countsplit.tutorials/" class="external-link">https://anna-neufeld.github.io/countsplit.tutorials/</a>. This change helps with overall package size and build time. Most of the tutorials currently make use of Poisson thinning, but we are in the process of adding more tutorials that use the negative binomial methodology.</p>
</div>
<div class="section level2">
<h2 id="how-can-i-get-countsplit">How can I get countsplit?<a class="anchor" aria-label="anchor" href="#how-can-i-get-countsplit"></a>
</h2>
<p>Countsplit 4.0.0 is now available on <code>CRAN</code>, and can be downloaded as follows.</p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"countsplit"</span><span class="op">)</span></span></code></pre>
<p>The latest development version can be downloaded from github as follows. In order to run this, you must first make sure that <code>remotes</code> is installed (and if it is not, run <code>install.packages("remotes")</code>).</p>
<pre><code><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"anna-neufeld/countsplit"</span>, ref <span class="op">=</span> <span class="st">'develop'</span><span class="op">)</span></span></code></pre>
<p>To also download the data needed to reproduce the package vignettes, be sure to also install the ``countsplit.tutorials” package. This must be downloaded from github (it will not be added to CRAN due to package size and the time it takes to build the vignettes).</p>
<pre><code><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"anna-neufeld/countsplit.tutorials"</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="where-can-i-learn-more">Where can I learn more?<a class="anchor" aria-label="anchor" href="#where-can-i-learn-more"></a>
</h2>
<p>Please visit our tutorial website <a href="https://anna-neufeld.github.io/countsplit.tutorials/" class="external-link">https://anna-neufeld.github.io/countsplit.tutorials/</a> to see an introduction to our framework on simple simulated data, as well as tutorials for integrating the count splitting package with common scRNA-seq analysis pipelines (Seurat, scran, and Monocle3).</p>
<p>Please visit <a href="https://github.com/anna-neufeld/countsplit_paper" class="external-link">https://github.com/anna-neufeld/countsplit_paper</a> for code to reproduce the figures and tables from our Poisson paper.</p>
<p>Please visit <a href="https://github.com/anna-neufeld/nbcs_paper_simulations" class="external-link">https://github.com/anna-neufeld/nbcs_paper_simulations</a> for code to reproduce the figures and tables from our negative binomial paper.</p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Neufeld, A.,Gao, L., Popp, J., Battle, A. &amp; Witten, D. (2022), ‘Inference after latent variable estimation for single-cell RNA sequencing data’, Biostatistics.</p>
<p>Neufeld, A.,Dharamshi, A., Gao, L., &amp; Witten, D. (2023), ‘Data thinning for convolution-closed distributions’, <a href="https://arxiv.org/abs/2301.07276/" class="external-link uri">https://arxiv.org/abs/2301.07276/</a> .</p>
<p>Neufeld, A., Popp, J., Gao, L., Battle, A. &amp; Witten, D. (2023), ‘Negative binomial count splitting for single-cell RNA sequencing data’. <a href="https://arxiv.org/pdf/2307.12985.pdf" class="external-link uri">https://arxiv.org/pdf/2307.12985.pdf</a> .</p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=countsplit" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/anna-neufeld/countsplit/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/anna-neufeld/countsplit/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing countsplit</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Anna Neufeld <br><small class="roles"> Author, maintainer, copyright holder </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Anna Neufeld.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
