<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tutorial: countsplitting and scran • countsplit</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Tutorial: countsplitting and scran">
<meta property="og:description" content="countsplit">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">countsplit</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/countsplit_tutorial.html">Getting started</a>
</li>
<li>
  <a href="../articles/scran_tutorial.html">Tutorial with scran</a>
</li>
<li>
  <a href="../articles/seurat_tutorial.html">Tutorial with Seurat</a>
</li>
<li>
  <a href="../articles/monocle3_tutorial.html">Tutorial with monocle3</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/anna-neufeld/countsplit/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tutorial: countsplitting and scran</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/anna-neufeld/countsplit/blob/HEAD/vignettes/scran_tutorial.Rmd" class="external-link"><code>vignettes/scran_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>scran_tutorial.Rmd</code></div>

    </div>

    
    
<p>Before using this tutorial, we recommend that you read through our ``introduction to count splitting” tutorial to understand our method in a simple package with simulated data.</p>
<p>In this tutorial, we use a real dataset from <span class="citation">(<strong>elorbany2022single?</strong>)</span> that is also used in our paper. The dataset contains 10,000 cells collected over 15 days of a directed differentiation protocol from induced pluripotent stem cells (IPSC) to cardiomyocytes (cm).</p>
<p>We start with a simple example where we use basic k-means clustering and Poisson GLMs on this data. We then move to a more complex example where we use functions that are built into the <code>scran</code> package.</p>
<div class="section level2">
<h2 id="install-scran">Install scran<a class="anchor" aria-label="anchor" href="#install-scran"></a>
</h2>
<p>If you don’t already have <code>scran</code>, you will need to run:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"scran"</span><span class="op">)</span></code></pre></div>
<p>Next, you should load the package, along with others that we will use in this tutorial.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/anna-neufeld/countsplit" class="external-link">countsplit</a></span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="simple-example-on-real-data-using-scran-">Simple example on real data using scran.<a class="anchor" aria-label="anchor" href="#simple-example-on-real-data-using-scran-"></a>
</h2>
<p>This data is included in this package as a SingleCellExperiment object, so it is simple to load. We will use a few steps from the <code>scran</code> package.</p>
<p>The main differences between this section and the last is that we will select a subset of highly variable genes and we will estimate and then account for size factors. We use the size factors estimated on the training set in our analysis! This follows our general principle of estimating all the preprocessing on the training set.</p>
<p>We first load the data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">cm</span><span class="op">)</span></code></pre></div>
<p>This is a single cell experiment object.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cm</span></code></pre></div>
<pre><code><span class="co">## class: SingleCellExperiment </span>
<span class="co">## dim: 21971 10000 </span>
<span class="co">## metadata(0):</span>
<span class="co">## assays(1): counts</span>
<span class="co">## rownames(21971): AL627309.1 AL627309.6 ... AC136352.4 AC007325.4</span>
<span class="co">## rowData names(0):</span>
<span class="co">## colnames(10000): E1_E1CD3col4_CATTTGTGCTTG E1_E1CD3col2_AGAATAAGTCAC</span>
<span class="co">##   ... E1_E1CD1col5_GTTACGCTAGTG E1_E1CD2col4_CCGCACAAGATC</span>
<span class="co">## colData names(26): orig.ident nCount_RNA ... pseudotime ident</span>
<span class="co">## reducedDimNames(0):</span>
<span class="co">## mainExpName: NULL</span>
<span class="co">## altExpNames(0):</span></code></pre>
<p>The main item in <code>cm</code> that we care about is the counts matrix, which contains 21,971 genes and 10000 cells. We can view a small subset of it now.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">cm</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 21971 10000</span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">cm</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## 10 x 10 sparse Matrix of class "dgCMatrix"</span></code></pre>
<pre><code><span class="co">##    [[ suppressing 10 column names 'E1_E1CD3col4_CATTTGTGCTTG', 'E1_E1CD3col2_AGAATAAGTCAC', 'E2_E2CD2col5_ATGAATGATGAA' ... ]]</span></code></pre>
<pre><code><span class="co">##                                 </span>
<span class="co">## AL627309.1  .  . . . . . . . . .</span>
<span class="co">## AL627309.6  .  . . . . . . . . .</span>
<span class="co">## AL627309.5  .  . . . . . . . . .</span>
<span class="co">## AL669831.3  .  . . . . . . . . .</span>
<span class="co">## MTND1P23    1 12 . . 3 1 . . 4 .</span>
<span class="co">## MTND2P28    6  3 . . . . 2 1 . .</span>
<span class="co">## MTCO1P12    2  7 2 . 2 . 1 2 3 .</span>
<span class="co">## MTCO2P12    .  1 . . . . . . . .</span>
<span class="co">## MTATP8P1    .  . . . . . . . . .</span>
<span class="co">## MTATP6P1   10  8 . . . . 3 . 2 .</span></code></pre>
<p>However, there is other important information included in this data. For example, the cells were collected from 19 individuals over the course of 15 days. This info is included in the object!! We don’t want to just extract the count matrices and ONLY work with those, because we don’t want to lose this info!</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">cm</span><span class="op">$</span><span class="va">individual</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">## 18520 18912 19093 18858 18508 18511 18907 18505 19190 18855 19159 18489 18517 </span>
<span class="co">##   472   760   925   732   581   701   362   438   524   283   326   400   534 </span>
<span class="co">## 18499 18870 19193 19209 19108 19127 </span>
<span class="co">##   771   700   373   554   211   353</span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">cm</span><span class="op">$</span><span class="va">diffday</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">##  day0  day1  day3  day5  day7 day11 day15 </span>
<span class="co">##  2383  2389  1691   967  1260   862   448</span></code></pre>
<p>We now extract the counts and perform count splitting. Note that these are all GENE BY CELL matrices- which is different than what we had before.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">cm</span><span class="op">)</span>
<span class="va">split</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/countsplit.html">countsplit</a></span><span class="op">(</span><span class="va">X</span>, epsilon<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span>
<span class="va">Xtrain</span> <span class="op">&lt;-</span> <span class="va">split</span><span class="op">$</span><span class="va">train</span>
<span class="va">Xtest</span> <span class="op">&lt;-</span> <span class="va">split</span><span class="op">$</span><span class="va">test</span></code></pre></div>
<p>Now we want to compute clusters on the training set. But this time, instead of simply running <code><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans()</a></code> on <code>log(Xtrain+1)</code>, we will use an existing scRNA-seq pipeline from the <code>scran</code> package that also involves preprocessing steps such as selecting highly variable genes. In order to do this, we need to do some analysis that expects a <code>SingleCellExperiment</code> object rather than a simple matrix. All we need to make the <code>SingleCellExperiment</code> training object is the <code>Xtrain</code> counts matrix. But it might be nice to retain the column data?</p>
<p>Actually in this case I don’t need any metadata for my analysis. So I could also just contruct a new scRNAseq object from the counts only.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cm.train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts<span class="op">=</span><span class="va">Xtrain</span><span class="op">)</span><span class="op">)</span>
<span class="fu">colData</span><span class="op">(</span><span class="va">cm.train</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">cm</span><span class="op">)</span></code></pre></div>
<p>Now we are ready for our analysis! These steps were inspired by the `<code>scran</code> tutorial, available at {<a href="https://bioconductor.org/packages/release/bioc/vignettes/scran/inst/doc/scran.html" class="external-link uri">https://bioconductor.org/packages/release/bioc/vignettes/scran/inst/doc/scran.html</a>}.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/quickCluster.html" class="external-link">quickCluster</a></span><span class="op">(</span><span class="va">cm.train</span><span class="op">)</span>
<span class="va">cm.train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html" class="external-link">computeSumFactors</a></span><span class="op">(</span><span class="va">cm.train</span>, clusters<span class="op">=</span><span class="va">clusters</span><span class="op">)</span>
<span class="va">cm.train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">cm.train</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">top.hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">cm.train</span><span class="op">)</span>, n<span class="op">=</span><span class="fl">2000</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cm.train</span><span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/fixedPCA.html" class="external-link">fixedPCA</a></span><span class="op">(</span><span class="va">cm.train</span>, subset.row<span class="op">=</span><span class="va">top.hvgs</span><span class="op">)</span>
<span class="va">clusters.train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/clusterCells.html" class="external-link">clusterCells</a></span><span class="op">(</span><span class="va">cm.train</span>,use.dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span></code></pre></div>
<p>It turns out that this function returned 11 clusters.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">clusters.train</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## clusters.train</span>
<span class="co">##    1    2    3    4    5    6    7 </span>
<span class="co">## 1997  698  588 2312 2983  958  464</span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">cm.train</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">PC1</span>, y<span class="op">=</span><span class="va">PC2</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">clusters.train</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>col<span class="op">=</span><span class="st">"Cluster"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Don't know how to automatically pick scale for object of type reduced.dim.matrix/matrix. Defaulting to continuous.</span>
<span class="co">## Don't know how to automatically pick scale for object of type reduced.dim.matrix/matrix. Defaulting to continuous.</span></code></pre>
<p><img src="scran_tutorial_files/figure-html/unnamed-chunk-12-1.png" width="90%"></p>
<p>We now consider two ways to check differential expression. The first is Poisson GLMs for differentially expressed genes between 1 and 2. We don’t need Xtest to be in a special object for this.For computational efficiency, we don’t want to check all 21,000 genes. Let’s check 500 randomly selected genes. Recall that in Xtest, the genes are the ROWs not the columns.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">indices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">clusters.train</span><span class="op">==</span><span class="fl">1</span> <span class="op">|</span> <span class="va">clusters.train</span><span class="op">==</span><span class="fl">2</span><span class="op">)</span>
<span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">NCOL</a></span><span class="op">(</span><span class="va">Xtest</span><span class="op">)</span>, size<span class="op">=</span><span class="fl">500</span><span class="op">)</span>
<span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">Xtest</span><span class="op">[</span><span class="va">genes</span>, <span class="va">indices</span><span class="op">]</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">u</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/glm.html" class="external-link">glm</a></span><span class="op">(</span><span class="va">u</span><span class="op">~</span><span class="va">clusters.train</span><span class="op">[</span><span class="va">indices</span><span class="op">]</span>, offset<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/sizeFactors.html" class="external-link">sizeFactors</a></span><span class="op">(</span><span class="va">cm.train</span><span class="op">)</span><span class="op">[</span><span class="va">indices</span><span class="op">]</span>, family<span class="op">=</span><span class="st">"poisson"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">results</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">0.01</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">## FALSE  TRUE </span>
<span class="co">##   428    72</span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##               Estimate Std. Error     z value    Pr(&gt;|z|)</span>
<span class="co">## FRRS1      -0.27893344  1.0953237 -0.25465845 0.798986910</span>
<span class="co">## SYNJ2      -0.01942228  0.4241392 -0.04579223 0.963475851</span>
<span class="co">## KCNMB2-AS1 -0.17357296  0.5527615 -0.31401061 0.753512984</span>
<span class="co">## TP53INP1   -0.51532223  0.6212037 -0.82955438 0.406790781</span>
<span class="co">## AC006970.1 -1.23444489  1.0376196 -1.18968921 0.234168568</span>
<span class="co">## C3orf62     1.42581453  0.4368847  3.26359421 0.001100086</span></code></pre>
<p>What if instead we try to use their own find markers function? It works, just need to figure out how to use. I think it does pairwise t-test between all of the things.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cm.test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts<span class="op">=</span><span class="va">Xtest</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/sizeFactors.html" class="external-link">sizeFactors</a></span><span class="op">(</span><span class="va">cm.test</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/sizeFactors.html" class="external-link">sizeFactors</a></span><span class="op">(</span><span class="va">cm.train</span><span class="op">)</span>
<span class="va">cm.test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">cm.test</span><span class="op">)</span></code></pre></div>
<p>This shows how different cluster 1 is from all of the other clusters. WRite a little paragaph about interpreting results. Ok! There really are important genes! Even without double dipping!!!</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu">scran</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scran/man/findMarkers.html" class="external-link">findMarkers</a></span><span class="op">(</span>
  <span class="va">cm.test</span>, groups<span class="op">=</span> <span class="va">clusters.train</span>,
  pval.type <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span>
<span class="va">results</span> <span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 21971 rows and 9 columns</span>
<span class="co">##               p.value         FDR summary.logFC     logFC.2     logFC.3</span>
<span class="co">##             &lt;numeric&gt;   &lt;numeric&gt;     &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;</span>
<span class="co">## TPM1      1.08718e-87 2.38865e-83      1.524155    -2.60820   -1.691339</span>
<span class="co">## TNNT2     3.92556e-20 3.42042e-16      0.243453    -2.74309   -0.629304</span>
<span class="co">## MYL7      4.67036e-20 3.42042e-16     -0.619474    -3.00084   -0.619474</span>
<span class="co">## NEXN      6.38871e-17 3.50916e-13      0.180813    -1.26727   -0.599001</span>
<span class="co">## ANKRD1    8.78375e-17 3.85975e-13      0.256620    -1.45896   -1.366888</span>
<span class="co">## ...               ...         ...           ...         ...         ...</span>
<span class="co">## GABRQ               1           1             0  0.00000000  0.00000000</span>
<span class="co">## CLIC2               1           1             0  0.00000000 -0.00355741</span>
<span class="co">## EEF1A1P41           1           1             0 -0.00491876  0.00000000</span>
<span class="co">## PCDH11Y             1           1             0  0.00000000  0.00000000</span>
<span class="co">## TMSB4Y              1           1             0  0.00000000  0.00000000</span>
<span class="co">##                logFC.4      logFC.5      logFC.6     logFC.7</span>
<span class="co">##              &lt;numeric&gt;    &lt;numeric&gt;    &lt;numeric&gt;   &lt;numeric&gt;</span>
<span class="co">## TPM1          1.393705     1.539964     1.764288    1.524155</span>
<span class="co">## TNNT2         0.261539     0.284571     0.299269    0.243453</span>
<span class="co">## MYL7          0.511705     0.595673     0.518480    0.422453</span>
<span class="co">## NEXN          0.165924     0.185703     0.184252    0.180813</span>
<span class="co">## ANKRD1        0.300527     0.322272     0.295123    0.256620</span>
<span class="co">## ...                ...          ...          ...         ...</span>
<span class="co">## GABRQ     -0.004987919 -0.002776015  0.000000000  0.00000000</span>
<span class="co">## CLIC2     -0.000899236  0.000000000 -0.002090804  0.00000000</span>
<span class="co">## EEF1A1P41 -0.000239274  0.000000000  0.000000000  0.00000000</span>
<span class="co">## PCDH11Y   -0.001452754 -0.001119289 -0.000813786  0.00000000</span>
<span class="co">## TMSB4Y     0.000000000 -0.000292628  0.000000000 -0.00585973</span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Anna Neufeld.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
