<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tutorial: countsplitting and scran • countsplit</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Tutorial: countsplitting and scran">
<meta property="og:description" content="countsplit">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">countsplit</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/countsplit_tutorial.html">Introductory tutorial</a>
</li>
<li>
  <a href="../articles/seurat_tutorial.html">Seurat tutorial</a>
</li>
<li>
  <a href="../articles/scran_tutorial.html">Scran tutorial</a>
</li>
<li>
  <a href="../articles/monocle3_tutorial.html">Monocle3 tutorial</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/anna-neufeld/countsplit/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tutorial: countsplitting and scran</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/anna-neufeld/countsplit/blob/HEAD/vignettes/scran_tutorial.Rmd" class="external-link"><code>vignettes/scran_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>scran_tutorial.Rmd</code></div>

    </div>

    
    
<p>Before using this tutorial, we recommend that you read through our <a href="https://anna-neufeld.github.io/countsplit/articles/countsplit_tutorial.html" class="external-link">introductory tutorial</a> to understand our method in a simple example with simulated data.</p>
<p>In this tutorial, we use a real dataset from <span class="citation">Elorbany et al. (2022)</span>. The dataset contains 10,000 cells collected over 15 days of a directed differentiation protocol from induced pluripotent stem cells (IPSC) to cardiomyocytes (cm).</p>
<div class="section level2">
<h2 id="install-scran">Install scran<a class="anchor" aria-label="anchor" href="#install-scran"></a>
</h2>
<p>If you don’t already have <code>scran</code>, you will need to run:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"scran"</span><span class="op">)</span></code></pre></div>
<p>Next, you should load the package, along with others that we will use in this tutorial.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/anna-neufeld/countsplit" class="external-link">countsplit</a></span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-the-data-and-perform-count-splitting">Load the data and perform count splitting<a class="anchor" aria-label="anchor" href="#load-the-data-and-perform-count-splitting"></a>
</h2>
<p>This data is included in this package as a <code>SingleCellExperiment</code> object, so it is simple to load.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">cm</span><span class="op">)</span>
<span class="va">cm</span></code></pre></div>
<pre><code><span class="co">## class: SingleCellExperiment </span>
<span class="co">## dim: 21971 10000 </span>
<span class="co">## metadata(0):</span>
<span class="co">## assays(1): counts</span>
<span class="co">## rownames(21971): AL627309.1 AL627309.6 ... AC136352.4 AC007325.4</span>
<span class="co">## rowData names(0):</span>
<span class="co">## colnames(10000): E1_E1CD3col4_CATTTGTGCTTG E1_E1CD3col2_AGAATAAGTCAC</span>
<span class="co">##   ... E1_E1CD1col5_GTTACGCTAGTG E1_E1CD2col4_CCGCACAAGATC</span>
<span class="co">## colData names(26): orig.ident nCount_RNA ... pseudotime ident</span>
<span class="co">## reducedDimNames(0):</span>
<span class="co">## mainExpName: NULL</span>
<span class="co">## altExpNames(0):</span></code></pre>
<p>The main item in <code>cm</code> that we care about is the counts matrix, which contains 21,971 genes and 10000 cells. We can view a small subset of it now.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">cm</span><span class="op">)</span><span class="op">)</span>
<span class="co">## [1] 21971 10000</span>
<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">cm</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>
<span class="co">## 10 x 10 sparse Matrix of class "dgCMatrix"</span>
<span class="co">##    [[ suppressing 10 column names 'E1_E1CD3col4_CATTTGTGCTTG', 'E1_E1CD3col2_AGAATAAGTCAC', 'E2_E2CD2col5_ATGAATGATGAA' ... ]]</span>
<span class="co">##                                 </span>
<span class="co">## AL627309.1  .  . . . . . . . . .</span>
<span class="co">## AL627309.6  .  . . . . . . . . .</span>
<span class="co">## AL627309.5  .  . . . . . . . . .</span>
<span class="co">## AL669831.3  .  . . . . . . . . .</span>
<span class="co">## MTND1P23    1 12 . . 3 1 . . 4 .</span>
<span class="co">## MTND2P28    6  3 . . . . 2 1 . .</span>
<span class="co">## MTCO1P12    2  7 2 . 2 . 1 2 3 .</span>
<span class="co">## MTCO2P12    .  1 . . . . . . . .</span>
<span class="co">## MTATP8P1    .  . . . . . . . . .</span>
<span class="co">## MTATP6P1   10  8 . . . . 3 . 2 .</span></code></pre></div>
<p>There is other important information included in the <code>cm</code> object. For example, the cells were collected from 19 individuals over the course of 15 days. Information on which individual and which day each cell is from is saved within the <code>cm</code> object as column data.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">cm</span><span class="op">$</span><span class="va">individual</span><span class="op">)</span>
<span class="co">## </span>
<span class="co">## 18520 18912 19093 18858 18508 18511 18907 18505 19190 18855 19159 18489 18517 </span>
<span class="co">##   472   760   925   732   581   701   362   438   524   283   326   400   534 </span>
<span class="co">## 18499 18870 19193 19209 19108 19127 </span>
<span class="co">##   771   700   373   554   211   353</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">cm</span><span class="op">$</span><span class="va">diffday</span><span class="op">)</span>
<span class="co">## </span>
<span class="co">##  day0  day1  day3  day5  day7 day11 day15 </span>
<span class="co">##  2383  2389  1691   967  1260   862   448</span></code></pre></div>
<p>To perform count splitting we wish to extract only the count matrix.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">cm</span><span class="op">)</span>
<span class="va">split</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/countsplit.html">countsplit</a></span><span class="op">(</span><span class="va">X</span>, epsilon<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span>
<span class="va">Xtrain</span> <span class="op">&lt;-</span> <span class="va">split</span><span class="op">$</span><span class="va">train</span>
<span class="va">Xtest</span> <span class="op">&lt;-</span> <span class="va">split</span><span class="op">$</span><span class="va">test</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="run-preprocessing-and-clustering-on-the-training-set-">Run preprocessing and clustering on the training set.<a class="anchor" aria-label="anchor" href="#run-preprocessing-and-clustering-on-the-training-set-"></a>
</h2>
<p>We now wish to compute clusters on the training set. Unlike in our <a href="https://anna-neufeld.github.io/countsplit/articles/countsplit_tutorial.html" class="external-link">introductory tuorial</a>, instead of simply running <code><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans()</a></code> on <code>log(Xtrain+1)</code>, we will use an existing scRNA-seq pipeline from the <code>scran</code> package that also involves preprocessing steps such as selecting highly variable genes. In order to do this, we need to store the training set count matrix <code>Xtrain</code> in a <code>SingleCellExperiment</code> object.</p>
<p>We could make a new <code>SingleCellExperiment</code> object that contains only the <code>Xtrain</code> counts matrix. However, this would discard all of the metadata that was stored in the <code>cm</code> object, such as information regarding which cells where collected on which days and from which individuals. Thus, we chose to let <code>cm.train</code> be a copy of <code>cm</code> where we only update the <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">counts()</a></code> attribute. This way, all metadata from <code>cm</code> is retained in <code>cm.train</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cm.train</span> <span class="op">&lt;-</span> <span class="va">cm</span>
<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">cm.train</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">Xtrain</span></code></pre></div>
<p>Now we are ready for our analysis. These steps were inspired by the <a href="https://bioconductor.org/packages/release/bioc/vignettes/scran/inst/doc/scran.html" class="external-link">scran tutorial</a>. We first compute sum factors and then we perform log normalization. Note that the <code><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html" class="external-link">computeSumFactors()</a></code> function requires computing initial clusters of cells using <code><a href="https://rdrr.io/pkg/scran/man/quickCluster.html" class="external-link">quickCluster()</a></code>; these are different from the clusters of cells that we will later analyze for differential expression.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/quickCluster.html" class="external-link">quickCluster</a></span><span class="op">(</span><span class="va">cm.train</span><span class="op">)</span>
<span class="va">cm.train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html" class="external-link">computeSumFactors</a></span><span class="op">(</span><span class="va">cm.train</span>, clusters<span class="op">=</span><span class="va">clusters</span><span class="op">)</span>
<span class="va">cm.train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">cm.train</span><span class="op">)</span></code></pre></div>
<p>We then compute the top 2000 highly variable genes. While this step does not alter the dimension of <code>counts(cm.train)</code>, it allows us to later perform clustering or dimension reduction using only these highly variable genes.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">top.hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">cm.train</span><span class="op">)</span>, n<span class="op">=</span><span class="fl">2000</span><span class="op">)</span></code></pre></div>
<p>Finally, we perform dimension reduction on the dataset (using only the highly variable genes), and then cluster the dimension-reduced dataset using scran’s <code>clusterCells</code> function (which performs graph-based clustering).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cm.train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/fixedPCA.html" class="external-link">fixedPCA</a></span><span class="op">(</span><span class="va">cm.train</span>, subset.row<span class="op">=</span><span class="va">top.hvgs</span><span class="op">)</span>
<span class="va">clusters.train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/clusterCells.html" class="external-link">clusterCells</a></span><span class="op">(</span><span class="va">cm.train</span>,use.dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span></code></pre></div>
<p>It turns out that this function returned 11 clusters. We can visualize them below.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">clusters.train</span><span class="op">)</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">cm.train</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">PC1</span>, y<span class="op">=</span><span class="va">PC2</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">clusters.train</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>col<span class="op">=</span><span class="st">"Cluster"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Don't know how to automatically pick scale for object of type reduced.dim.matrix/matrix. Defaulting to continuous.</span>
<span class="co">## Don't know how to automatically pick scale for object of type reduced.dim.matrix/matrix. Defaulting to continuous.</span></code></pre>
<p><img src="scran_tutorial_files/figure-html/unnamed-chunk-11-1.png" width="90%"></p>
</div>
<div class="section level2">
<h2 id="differential-expression-testing-with-poisson-glms">Differential expression testing with Poisson GLMs<a class="anchor" aria-label="anchor" href="#differential-expression-testing-with-poisson-glms"></a>
</h2>
<p>We now consider fitting Poisson GLMs to test for differential expression between clusters 1 and 2. We can do this analysis “by hand,” meaning that we do not need to store <code>Xtest</code> in a <code>SingleCellExperiment</code> object.</p>
<p>For computational efficiency, we test 500 randomly selected genes for differential expression, rather than checking all 21,000 genes. The first step is to subset the test data to only include these 500 genes and to only include cells that were placed in cluster 1 or cluster 2. We will include size factors for each cell, estimated on the training set, as offsets in our Poisson GLM. We need to obtain the appropriate size factor vector and subset it to only include the cells that were placed in cluster 1 or cluster 2. We do these steps below.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">indices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">clusters.train</span><span class="op">==</span><span class="fl">1</span> <span class="op">|</span> <span class="va">clusters.train</span><span class="op">==</span><span class="fl">2</span><span class="op">)</span>
<span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">NCOL</a></span><span class="op">(</span><span class="va">Xtest</span><span class="op">)</span>, size<span class="op">=</span><span class="fl">500</span><span class="op">)</span> 
<span class="va">Xtestsubset</span> <span class="op">&lt;-</span> <span class="va">Xtest</span><span class="op">[</span><span class="va">genes</span>, <span class="va">indices</span><span class="op">]</span>
<span class="va">sizefactors.subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/sizeFactors.html" class="external-link">sizeFactors</a></span><span class="op">(</span><span class="va">cm.train</span><span class="op">)</span><span class="op">[</span><span class="va">indices</span><span class="op">]</span></code></pre></div>
<p>We are now ready to test for differential expression. We see that 61 out of our 500 genes were identified as significantly differentially expressed at alpha=0.01.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">Xtestsubset</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">u</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/glm.html" class="external-link">glm</a></span><span class="op">(</span><span class="va">u</span><span class="op">~</span><span class="va">clusters.train</span><span class="op">[</span><span class="va">indices</span><span class="op">]</span>, offset<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">sizefactors.subset</span><span class="op">)</span>, family<span class="op">=</span><span class="st">"poisson"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">results</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">0.01</span><span class="op">)</span> 
<span class="co">## </span>
<span class="co">## FALSE  TRUE </span>
<span class="co">##   439    61</span></code></pre></div>
<p>We can view the names of these highly significant genes below.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">results</span><span class="op">[</span><span class="va">results</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">0.01</span>,<span class="op">]</span>, n<span class="op">=</span><span class="fl">10</span><span class="op">)</span>
<span class="co">##           Estimate Std. Error   z value     Pr(&gt;|z|)</span>
<span class="co">## C3orf62  1.2324039 0.43690137  2.820783 4.790661e-03</span>
<span class="co">## PPFIA4   1.6967096 0.62676308  2.707099 6.787406e-03</span>
<span class="co">## CDCA7L   0.6235322 0.14797067  4.213890 2.510091e-05</span>
<span class="co">## ZFR      0.2439864 0.08567689  2.847750 4.402946e-03</span>
<span class="co">## SELENOT  0.5564246 0.18215736  3.054637 2.253332e-03</span>
<span class="co">## COL11A2  2.9288532 1.08012328  2.711592 6.696103e-03</span>
<span class="co">## FAM241A  1.6967096 0.62676433  2.707093 6.787516e-03</span>
<span class="co">## PPP2R5A  0.5493071 0.21081767  2.605603 9.171275e-03</span>
<span class="co">## ZNF138  -0.8594601 0.32129460 -2.674991 7.473132e-03</span>
<span class="co">## ZNF436   1.6070974 0.57008420  2.819053 4.816561e-03</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="differential-expression-testing-with-scran">Differential expression testing with scran<a class="anchor" aria-label="anchor" href="#differential-expression-testing-with-scran"></a>
</h2>
<p>In this section, instead of testing for differential expression “by hand” using <code><a href="https://rdrr.io/r/stats/glm.html" class="external-link">glm()</a></code>, we use the <code><a href="https://rdrr.io/pkg/scran/man/scoreMarkers.html" class="external-link">scoreMarkers()</a></code> function from the scran package. This is the method used in the <a href="https://bioconductor.org/packages/release/bioc/vignettes/scran/inst/doc/scran.html#6_Identifying_marker_genes" class="external-link">scran tutorial</a>, in the section titled “Find Marker Genes.”</p>
<p>In order to use the <code><a href="https://rdrr.io/pkg/scran/man/scoreMarkers.html" class="external-link">scoreMarkers()</a></code> function, we need to store our test matrix in a <code>SingleCellExperiment</code> object. We can either construct this from scratch using only the count matrix, or we could make a copy of the original <code>cm</code> object and add the count matrix after. Since we do not anticipate needing any metadata in this section, we construct the object from scratch.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cm.test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts<span class="op">=</span><span class="va">Xtest</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The <code><a href="https://rdrr.io/pkg/scran/man/scoreMarkers.html" class="external-link">scoreMarkers()</a></code> function needs access to the log-normalized counts assay (<code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts()</a></code>) in the <code>cm.test</code> object. To fill in this assay, we need to run the <code>logNormCounts</code> function on <code>cm.test</code>. We would like to normalize the test matrix using size factors computed on the training set (this follows our general philosophy of performing all preprocessing steps on the training set). Thus, we obtain size factors from the training set before calling <code><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts()</a></code>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/sizeFactors.html" class="external-link">sizeFactors</a></span><span class="op">(</span><span class="va">cm.test</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/sizeFactors.html" class="external-link">sizeFactors</a></span><span class="op">(</span><span class="va">cm.train</span><span class="op">)</span>
<span class="va">cm.test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">cm.test</span><span class="op">)</span> </code></pre></div>
<p>This first element in <code>results</code> shows marker genes that distinguish cluster 1 from all other clusters.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu">scran</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scran/man/findMarkers.html" class="external-link">findMarkers</a></span><span class="op">(</span> 
<span class="va">cm.test</span>, groups<span class="op">=</span> <span class="va">clusters.train</span>,
 pval.type <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span> 
<span class="va">results</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 21971 rows and 9 columns</span>
<span class="co">##               p.value         FDR summary.logFC     logFC.2     logFC.3</span>
<span class="co">##             &lt;numeric&gt;   &lt;numeric&gt;     &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;</span>
<span class="co">## TPM1      1.08718e-87 2.38865e-83      1.524155    -2.60820   -1.691339</span>
<span class="co">## TNNT2     3.92556e-20 3.42042e-16      0.243453    -2.74309   -0.629304</span>
<span class="co">## MYL7      4.67036e-20 3.42042e-16     -0.619474    -3.00084   -0.619474</span>
<span class="co">## NEXN      6.38871e-17 3.50916e-13      0.180813    -1.26727   -0.599001</span>
<span class="co">## ANKRD1    8.78375e-17 3.85975e-13      0.256620    -1.45896   -1.366888</span>
<span class="co">## ...               ...         ...           ...         ...         ...</span>
<span class="co">## GABRQ               1           1             0  0.00000000  0.00000000</span>
<span class="co">## CLIC2               1           1             0  0.00000000 -0.00355741</span>
<span class="co">## EEF1A1P41           1           1             0 -0.00491876  0.00000000</span>
<span class="co">## PCDH11Y             1           1             0  0.00000000  0.00000000</span>
<span class="co">## TMSB4Y              1           1             0  0.00000000  0.00000000</span>
<span class="co">##                logFC.4      logFC.5      logFC.6     logFC.7</span>
<span class="co">##              &lt;numeric&gt;    &lt;numeric&gt;    &lt;numeric&gt;   &lt;numeric&gt;</span>
<span class="co">## TPM1          1.393705     1.539964     1.764288    1.524155</span>
<span class="co">## TNNT2         0.261539     0.284571     0.299269    0.243453</span>
<span class="co">## MYL7          0.511705     0.595673     0.518480    0.422453</span>
<span class="co">## NEXN          0.165924     0.185703     0.184252    0.180813</span>
<span class="co">## ANKRD1        0.300527     0.322272     0.295123    0.256620</span>
<span class="co">## ...                ...          ...          ...         ...</span>
<span class="co">## GABRQ     -0.004987919 -0.002776015  0.000000000  0.00000000</span>
<span class="co">## CLIC2     -0.000899236  0.000000000 -0.002090804  0.00000000</span>
<span class="co">## EEF1A1P41 -0.000239274  0.000000000  0.000000000  0.00000000</span>
<span class="co">## PCDH11Y   -0.001452754 -0.001119289 -0.000813786  0.00000000</span>
<span class="co">## TMSB4Y     0.000000000 -0.000292628  0.000000000 -0.00585973</span></code></pre>
<p>Even with count splitting, we identify several highly significant marker genes, both “by hand” and with the <code><a href="https://rdrr.io/pkg/scran/man/findMarkers.html" class="external-link">findMarkers()</a></code> function.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-elorbany2022single" class="csl-entry">
Elorbany, Reem, Joshua M Popp, Katherine Rhodes, Benjamin J Strober, Kenneth Barr, Guanghao Qi, Yoav Gilad, and Alexis Battle. 2022. <span>“Single-Cell Sequencing Reveals Lineage-Specific Dynamic Genetic Regulation of Gene Expression During Human Cardiomyocyte Differentiation.”</span> <em>PLoS Genetics</em> 18 (1): e1009666.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Anna Neufeld.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.4.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
