<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tutorial: countsplitting and scran • countsplit</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Tutorial: countsplitting and scran">
<meta property="og:description" content="countsplit">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">countsplit</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/countsplit_tutorial.html">Introductory tutorial</a>
</li>
<li>
  <a href="../articles/seurat_tutorial.html">Seurat tutorial</a>
</li>
<li>
  <a href="../articles/scran_tutorial.html">Scran tutorial</a>
</li>
<li>
  <a href="../articles/monocle3_tutorial.html">Monocle3 tutorial</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/anna-neufeld/countsplit/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tutorial: countsplitting and scran</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/anna-neufeld/countsplit/blob/HEAD/vignettes/scran_tutorial.Rmd" class="external-link"><code>vignettes/scran_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>scran_tutorial.Rmd</code></div>

    </div>

    
    
<p>Before using this tutorial, we recommend that you read through our <a href="https://anna-neufeld.github.io/countsplit/articles/countsplit_tutorial.html" class="external-link">introductory tuorial</a> to understand our method in a simple example with simulated data.</p>
<p>In this tutorial, we use a real dataset from <span class="citation">Elorbany et al. (2022)</span>. The dataset contains 10,000 cells collected over 15 days of a directed differentiation protocol from induced pluripotent stem cells (IPSC) to cardiomyocytes (cm).</p>
<div class="section level2">
<h2 id="install-scran">Install scran<a class="anchor" aria-label="anchor" href="#install-scran"></a>
</h2>
<p>If you don’t already have <code>scran</code>, you will need to run:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"scran"</span><span class="op">)</span></code></pre></div>
<p>Next, you should load the package, along with others that we will use in this tutorial.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/anna-neufeld/countsplit" class="external-link">countsplit</a></span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-the-data-and-perform-count-splitting">Load the data and perform count splitting<a class="anchor" aria-label="anchor" href="#load-the-data-and-perform-count-splitting"></a>
</h2>
<p>This data is included in this package as a <code>SingleCellExperiment</code> object, so it is simple to load.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">cm</span><span class="op">)</span>
<span class="va">cm</span></code></pre></div>
<pre><code><span class="co">## class: SingleCellExperiment </span>
<span class="co">## dim: 21971 10000 </span>
<span class="co">## metadata(0):</span>
<span class="co">## assays(1): counts</span>
<span class="co">## rownames(21971): AL627309.1 AL627309.6 ... AC136352.4 AC007325.4</span>
<span class="co">## rowData names(0):</span>
<span class="co">## colnames(10000): E1_E1CD3col4_CATTTGTGCTTG E1_E1CD3col2_AGAATAAGTCAC</span>
<span class="co">##   ... E1_E1CD1col5_GTTACGCTAGTG E1_E1CD2col4_CCGCACAAGATC</span>
<span class="co">## colData names(26): orig.ident nCount_RNA ... pseudotime ident</span>
<span class="co">## reducedDimNames(0):</span>
<span class="co">## mainExpName: NULL</span>
<span class="co">## altExpNames(0):</span></code></pre>
<p>The main item in <code>cm</code> that we care about is the counts matrix, which contains 21,971 genes and 10000 cells. We can view a small subset of it now.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">cm</span><span class="op">)</span><span class="op">)</span>
<span class="co">## [1] 21971 10000</span>
<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">cm</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>
<span class="co">## 10 x 10 sparse Matrix of class "dgCMatrix"</span>
<span class="co">##    [[ suppressing 10 column names 'E1_E1CD3col4_CATTTGTGCTTG', 'E1_E1CD3col2_AGAATAAGTCAC', 'E2_E2CD2col5_ATGAATGATGAA' ... ]]</span>
<span class="co">##                                 </span>
<span class="co">## AL627309.1  .  . . . . . . . . .</span>
<span class="co">## AL627309.6  .  . . . . . . . . .</span>
<span class="co">## AL627309.5  .  . . . . . . . . .</span>
<span class="co">## AL669831.3  .  . . . . . . . . .</span>
<span class="co">## MTND1P23    1 12 . . 3 1 . . 4 .</span>
<span class="co">## MTND2P28    6  3 . . . . 2 1 . .</span>
<span class="co">## MTCO1P12    2  7 2 . 2 . 1 2 3 .</span>
<span class="co">## MTCO2P12    .  1 . . . . . . . .</span>
<span class="co">## MTATP8P1    .  . . . . . . . . .</span>
<span class="co">## MTATP6P1   10  8 . . . . 3 . 2 .</span></code></pre></div>
<p>There is other important information included in the <code>cm</code> object. For example, the cells were collected from 19 individuals over the course of 15 days. Information on which individual and which day each cell is from is saved within the <code>cm</code> object as column data.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">cm</span><span class="op">$</span><span class="va">individual</span><span class="op">)</span>
<span class="co">## </span>
<span class="co">## 18520 18912 19093 18858 18508 18511 18907 18505 19190 18855 19159 18489 18517 </span>
<span class="co">##   472   760   925   732   581   701   362   438   524   283   326   400   534 </span>
<span class="co">## 18499 18870 19193 19209 19108 19127 </span>
<span class="co">##   771   700   373   554   211   353</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">cm</span><span class="op">$</span><span class="va">diffday</span><span class="op">)</span>
<span class="co">## </span>
<span class="co">##  day0  day1  day3  day5  day7 day11 day15 </span>
<span class="co">##  2383  2389  1691   967  1260   862   448</span></code></pre></div>
<p>To perform count splitting we wish to extract only the count matrix.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">cm</span><span class="op">)</span>
<span class="va">split</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/countsplit.html">countsplit</a></span><span class="op">(</span><span class="va">X</span>, epsilon<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span>
<span class="va">Xtrain</span> <span class="op">&lt;-</span> <span class="va">split</span><span class="op">$</span><span class="va">train</span>
<span class="va">Xtest</span> <span class="op">&lt;-</span> <span class="va">split</span><span class="op">$</span><span class="va">test</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="run-preprocessing-and-clustering-on-the-training-set-">Run preprocessing and clustering on the training set.<a class="anchor" aria-label="anchor" href="#run-preprocessing-and-clustering-on-the-training-set-"></a>
</h2>
<p>We now wish to compute clusters on the training set. Unlike in our [introductory tutorial]</p>
<p>Now we want to compute clusters on the training set. But this time, instead of simply running <code><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans()</a></code> on <code>log(Xtrain+1)</code>, we will use an existing scRNA-seq pipeline from the <code>scran</code> package that also involves preprocessing steps such as selecting highly variable genes. In order to do this, we need to do some analysis that expects a <code>SingleCellExperiment</code> object rather than a simple matrix.</p>
<p>All we need to make the <code>SingleCellExperiment</code> training object is the <code>Xtrain</code> counts matrix. But it might be nice to retain the metadata in case we need it later. In this case, we do that by copying the original <code>cm</code> object and then updating the <code>counts</code> assay.</p>
<p>Actually in this case I don’t need any metadata for my analysis. So I could also just contruct a new scRNAseq object from the counts only.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cm.train</span> <span class="op">&lt;-</span> <span class="va">cm</span>
<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">cm.train</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">Xtrain</span></code></pre></div>
<!-- Now we are ready for our analysis! These steps were inspired by the [`\textt{scran}``scran`` tutorial](https://bioconductor.org/packages/release/bioc/vignettes/scran/inst/doc/scran.html).  -->
<!-- We first compute sum factors for normalization and then we perform log normalization of the dataset.  -->
<!-- ```{r} -->
<!-- clusters <- quickCluster(cm.train) -->
<!-- cm.train <- computeSumFactors(cm.train, clusters=clusters) -->
<!-- cm.train <- logNormCounts(cm.train) -->
<!-- ``` -->
<!-- We then recide to work only with the top 2000 highly variable features for the remainder of the analysis. As noted in the ``scran`` tutorial, rather than select the top 2000 we could also try to use something like a FDR threshold to select only those that seem significantly variable.  -->
<!-- ```{r} -->
<!-- top.hvgs <- getTopHVGs(modelGeneVar(cm.train), n=2000) -->
<!-- ``` -->
<!-- Finally, we run scran's ``clusterCells`` function, which is a wrapper function for the graph-based clustering steps carried out in the tutorial.  -->
<!-- ```{r} -->
<!-- cm.train<- fixedPCA(cm.train, subset.row=top.hvgs) -->
<!-- clusters.train <- clusterCells(cm.train,use.dimred="PCA") -->
<!-- ``` -->
<!-- It turns out that this function returned 11 clusters. We can visualize them below.  -->
<!-- ```{r,out.width="90%"} -->
<!-- table(clusters.train) -->
<!-- ggplot(as_tibble(reducedDim(cm.train)), aes(x=PC1, y=PC2, col=as.factor(clusters.train)))+geom_point()+labs(col="Cluster") -->
<!-- ``` -->
<!-- # Differential expression testing with Poisson GLMs -->
<!-- We now consider two ways to check differential expression. The first is Poisson GLMs for differentially expressed genes between clusters 1 and 2. We don't need ``Xtest`` to be in a special object for this. For computational efficiency, we don't want to check all 21,000 genes. Let's check 500 randomly selected genes. -->
<!-- Recall that in ``Xtest``, the genes are the rows and not the columns. As in our manuscript, we include size factors as offsets here.  -->
<!-- ```{r, warning=FALSE} -->
<!-- set.seed(1) -->
<!-- indices <- which(clusters.train==1 | clusters.train==2) -->
<!-- genes <- sample(1:NCOL(Xtest), size=500) -->
<!-- results <- t(apply(Xtest[genes, indices], 1, function(u) summary(glm(u~clusters.train[indices], offset=sizeFactors(cm.train)[indices], family="poisson"))$coefficients[2,])) -->
<!-- table(results[,4] < 0.01) -->
<!-- head(results) -->
<!-- ``` -->
<!-- # Differential expression testing with scran -->
<!-- If instead of using our own ``glm`` code we want to use the ``scoreMarkers`` function from the ``scran`` package, as in the scran tutorial, we need to store our test matrix in a ``SingleCellExperiment`` object.  -->
<!-- We can either construct this from scratch using only the count matrix, or we could make a copy of the original ``cm`` object and add the count matrix after. It depends on if we want to be sure to obtain the metadata or not.  -->
<!-- ```{r} -->
<!-- cm.test <- SingleCellExperiment(list(counts=Xtest)) -->
<!-- sizeFactors(cm.test) <- sizeFactors(cm.train) -->
<!-- cm.test <- logNormCounts(cm.test) -->
<!-- ``` -->
<!-- This first element in ``results`` shows marker genes that distinguish cluster 1 from all other clusters. We see that, even with count splitting, there are highly significant marker genes. -->
<!-- ```{r} -->
<!-- results <- scran::findMarkers( -->
<!--   cm.test, groups= clusters.train, -->
<!--   pval.type = "all") -->
<!-- results[[1]] -->
<!-- ``` -->
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-elorbany2022single" class="csl-entry">
Elorbany, Reem, Joshua M Popp, Katherine Rhodes, Benjamin J Strober, Kenneth Barr, Guanghao Qi, Yoav Gilad, and Alexis Battle. 2022. <span>“Single-Cell Sequencing Reveals Lineage-Specific Dynamic Genetic Regulation of Gene Expression During Human Cardiomyocyte Differentiation.”</span> <em>PLoS Genetics</em> 18 (1): e1009666.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Anna Neufeld.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.4.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
